lib/DxcSupport/HLSLOptions.cpp

```diff
@@ -492,12 +492,28 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
   if (!limit.empty())
     opts.ScanLimit = std::stoul(std::string(limit));
 
-  opts.DxcOptimizationOptions = {};
-  std::vector<std::string> DisabledOptimizations = Args.getAllArgValues(OPT_opt_disable);
-  for (std::string opt : DisabledOptimizations) {
-    llvm::StringRef gvn("gvn");
-    if (gvn.equals_lower(opt))
-      opts.DxcOptimizationOptions.DisableGVN = true;
+  for (std::string opt : Args.getAllArgValues(OPT_opt_disable))
+    opts.DxcOptimizationToggles[llvm::StringRef(opt).upper()] = false;
+
+  for (std::string opt : Args.getAllArgValues(OPT_opt_enable)) {
+    if (!opts.DxcOptimizationToggles.insert ( {llvm::StringRef(opt).upper(), true} ).second) {
+      errors << "Contradictory use of -opt-disable and -opt-enable with "
+             << llvm::StringRef(opt).upper();
+      return 1;
+    }
+  }
+
+  std::vector<std::string> optSelects = Args.getAllArgValues(OPT_opt_select);
+  for (unsigned i = 0; i + 1 < optSelects.size(); i+=2) {
+    llvm::StringRef optimization = optSelects[i];
+    llvm::StringRef selection = optSelects[i+1];
+    if (opts.DxcOptimizationSelects.count(optimization.upper()) &&
+        selection.upper().compare(opts.DxcOptimizationSelects[optimization.upper()])) {
+      errors << "Contradictory -opt-selects for "
+             << optimization.upper();
+      return 1;
+    }
+    opts.DxcOptimizationSelects[optimization.upper()] = selection.upper();
```

Comment:
It was to prevent two values that only differ in case. Perhaps that doesn't make sense for these though. It would be weird for valid options to only differ by case, but this forces them to all be uppercase, which is worse than allowing the risk of a user capitalization mistake failing to enable a selection.